src/services/swapIntegrators/swapProviderParallelExecutor.ts

```diff
@@ -36,16 +37,61 @@ export class SwapProviderParallelExecutor {
   }
 
   async #fetchFromAll<T>(fetchMethod: (provider: SwapProvider) => Promise<T | Error>): Promise<T> {
-    const apiResponses = await Promise.all(
-      this.#providers.map((provider: SwapProvider) => fetchMethod(provider))
+    const MIN_WAIT = 3000 // 3s
+    const MAX_WAIT_AFTER_FIRST_COMPLETED = 2000 // 2s
+    const MAX_ABSOLUTE_WAIT_FOR_ALL_TO_COMPLETE = 15000 // 15s
+
+    const results: { provider: SwapProvider; result: T | Error }[] = []
+
+    const tasks = this.#providers.map((provider) =>
+      fetchMethod(provider)
+        .then((result) => ({ provider, result }))
+        .catch((err) => ({ provider, result: err as Error }))
     )
-    const resultsWithoutErrors = apiResponses.filter((r) => !(r instanceof Error))
-    if (resultsWithoutErrors.length) return resultsWithoutErrors.flat() as T
 
-    const errors = apiResponses.filter((r): r is Error => r instanceof Error)
+    const absoluteTimeout = wait(MAX_ABSOLUTE_WAIT_FOR_ALL_TO_COMPLETE).then(() => {
+      throw new Error(
+        'Our service providers are temporarily unavailable or your internet connection is too slow.'
+      )
+    })
+
+    const startTime = Date.now()
+    const firstResult = await Promise.race([Promise.any(tasks), absoluteTimeout])
+
+    if ('provider' in firstResult && 'result' in firstResult) {
+      results.push(firstResult)
+    }
+
+    const remainingTasks = this.#providers
+      .filter((p) => !results.some((r) => r.provider === p))
+      .map((provider, idx) =>
+        tasks[idx].then((res) => res).catch((err) => ({ provider, result: err as Error }))
+      )
+
+    // Figure out how long we've already waited
+    const elapsed = Date.now() - startTime
+    // If first was too quick, extend wait time so total ≥ MIN_WAIT
+    const remainingMinWait = Math.max(0, MIN_WAIT - elapsed)
+
+    const secondResult = (await Promise.race([
+      Promise.any(remainingTasks),
+      wait(MAX_WAIT_AFTER_FIRST_COMPLETED + remainingMinWait)
+    ])) as { provider: SwapProvider; result: Error | T }
```

Comment:
We must include `absoluteTimeout` in the race. If the first provider takes 14 sec, then this can exceed the timeout by going to 16.